packages/vm/src/imports.rs

```diff
@@ -533,6 +533,48 @@ pub fn do_db_next<A: BackendApi + 'static, S: Storage + 'static, Q: Querier + 's
     write_to_contract(data, &mut store, &out_data)
 }
 
+#[cfg(feature = "iterator")]
+pub fn do_db_next_key<A: BackendApi + 'static, S: Storage + 'static, Q: Querier + 'static>(
+    mut env: FunctionEnvMut<Environment<A, S, Q>>,
+    iterator_id: u32,
+) -> VmResult<u32> {
+    let (data, mut store) = env.data_and_store_mut();
+
+    let (result, gas_info) =
+        data.with_storage_from_context::<_, _>(|store| Ok(store.next_key(iterator_id)))?;
+
+    process_gas_info(data, &mut store, gas_info)?;
+
+    // Empty key will later be treated as _no more element_.
+    let key = result?.unwrap_or_default();
+
+    write_to_contract(data, &mut store, &key)
+}
+
+#[cfg(feature = "iterator")]
+pub fn do_db_next_value<A: BackendApi + 'static, S: Storage + 'static, Q: Querier + 'static>(
+    mut env: FunctionEnvMut<Environment<A, S, Q>>,
+    iterator_id: u32,
+) -> VmResult<u32> {
+    let (data, mut store) = env.data_and_store_mut();
+
+    let (result, gas_info) =
+        data.with_storage_from_context::<_, _>(|store| Ok(store.next_value(iterator_id)))?;
+
+    process_gas_info(data, &mut store, gas_info)?;
+
+    // Empty key will later be treated as _no more element_.
+    // No need to encode this further, as it is a single value.
+    let (mut value, has_next) = match result? {
```

Comment:
The term "has_next" was not obvious for me to understand. It could be interpreted as the element following the current value. But it's really an optional encoding of the current iteration. Maybe this is easier to audit:

```suggestion
    let (has_value, mut value) = match result? {
```

Reply:
I agree the `has_next` is not good. It was based on a misunderstanding I had while reading the `db_next` description.
We don't need this variable anyways if we change it to return a null pointer.